packages/datadog-plugin-sharedb/src/index.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+const MessagesAwaitingResponse = new WeakMap();
+const READABLE_ACTION_NAMES = {
+  hs: 'handshake',
+  qf: 'query-fetch',
+  qs: 'query-subscribe',
+  qu: 'query-unsubscribe',
+  bf: 'bulk-fetch',
+  bs: 'bulk-subscribe',
+  bu: 'bulk-unsubscribe',
+  f: 'fetch',
+  s: 'subscribe',
+  u: 'unsubscribe',
+  op: 'op',
+  nf: 'snapshot-fetch',
+  nt: 'snapshot-fetch-by-ts',
+  p: 'presence-broadcast',
+  pr: 'presence-request',
+  ps: 'presence-subscribe',
+  pu: 'presence-unsubscribe'
+};
+
+function createTraceName(action, collection) {
+  let actionName = READABLE_ACTION_NAMES[action];
+  if (actionName === undefined) {
+    actionName = action;
+  }
+  let traceName = 'sharedb-request/' + actionName;
+  if (collection) {
+    traceName += '/' + collection;
+  }
+  return traceName;
+}
+
+function createWrapHandle(tracer, config) { // called once
+  return function wrapTrigger(triggerFn) { // called once
+    return function handleMessageWithTrace(action, agent, triggerContext, callback) { // called for each trigger
+      /**
+       * What we're doing here is tying ourselves into the ShareDB Backend middleware.
+       * This allows us to create traces for all events that have triggers, like receiving a message and replying
+       * to it. The benefit of doing this over wrapping the connection class is that both the receive and reply
+       * triggers have access to a reference of the original request object. This allows us to use a WeakMap to
+       * store the span call backs to help prevent memory leaks.
+       *
+       */
+      switch (action) {
+        case 'receive':
+          if (triggerContext.data && triggerContext.data.a) {
+            const scope = tracer.scope();
+            const childOf = scope.active();
+            return triggerFn.call(this, action, agent, triggerContext, function wrappedCallback(err) {
```

Comment:
Could `tracer.wrap` be used instead here? It wraps callbacks automatically which might make this code a bit simpler.

Reply:
Attempted tracer.wrap, haven't gotten it to work yet: https://github.com/DataDog/dd-trace-js/commit/3192f67ebc281810fe082fff916d36c43f85dbcb